projects/utils-maps/src/lib/ol/utils-ol-mapstate.ts

```diff
@@ -0,0 +1,204 @@
+    // upon it
+    return;
+  }
+  const delta = 1, duration = 250;
+  const currentZoom = view.getZoom();
+  if (currentZoom !== undefined) {
+    const newZoom = view.getConstrainedZoom(currentZoom + delta);
+    if (duration > 0) {
+      if (view.getAnimating()) {
+        view.cancelAnimations();
+      }
+      view.animate({
+        zoom: newZoom,
+        duration,
+        easing: easeOut
+      });
+    } else {
+      view.setZoom(newZoom);
+    }
+  }
+}
+
+
+
+export function getMapProjection(map: Map) {
+  return map.getView().getProjection();
+}
+
+
+/**
+ * vector layers will be reprojected automatically
+ * wms layers will be updated with corresponding proj def in the requests.
+ * for other raster layers and for those wms layers whose backend does not support target projection, please
+ * define initial(default) layer projection, so openlayers will reproject on the client side
+ * projection is proj~ProjectionLike
+ */
+export function setMapProjection(map: Map, projection: Projection | string, viewOptions?: ViewOptions, cb?: (projection: Projection) => void) {
+  let epsg = DEFAULT_PROJECTION;
+  let view = null;
+  if (projection) {
+    let newViewOptions: ViewOptions = {};
+    if (viewOptions) {
+      newViewOptions = viewOptions;
+      newViewOptions.minResolution = undefined;
+      newViewOptions.maxResolution = undefined;
+      newViewOptions.resolution = undefined;
+      newViewOptions.resolutions = undefined;
+    }
+    if (projection instanceof Projection) {
+      newViewOptions.projection = projection;
+      const newCenter = transform(map.getView().getCenter(), map.getView().getProjection(), projection); // get center coordinates in the new projection
+      newViewOptions.center = newCenter; // map.getView().getCenter();
+      // _viewOptions.extent = projection.getExtent();// || undefined;
+      newViewOptions.zoom = map.getView().getZoom();
+    } else if (typeof projection === 'string') {
+      newViewOptions.projection = projection;
+      newViewOptions.center = map.getView().getCenter();
+      newViewOptions.zoom = map.getView().getZoom();
+    }
+    const mapView = new View(newViewOptions);
+    const oldProjection = DEFAULT_PROJECTION;
+    epsg = mapView.getProjection().getCode();
+    map.setView(mapView);
+    view = map.getView();
+
+    // reprojecting vector layers
+    map.getLayers().getArray().forEach((layerGroup: LayerGroup) => {
+      layerGroup.getLayers().getArray().forEach(layer => {
+        if (layer instanceof Layer) {
+          let source = layer.getSource();
+          // check for nested sources, e.g. cluster or cluster of clusters etc
+          while (source['source']) {
+            source = source['source'];
+          }
+          if (source instanceof VectorSource) {
+            reprojectVectorSource(source, oldProjection, DEFAULT_PROJECTION);
+          }
+        }
+      });
+    });
+    cb(getMapProjection(map));
```

Comment:
I think this callback-function `cb` serves no purpose here. If the user wants to do something with the new projection, he/she can just do that once the call to `setMapProjection` has finished. This is a synchronous function, after all - so we don't need to wait for a later point in time where `cb` could be called. Am I right?